examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c

```diff
@@ -130,469 +138,565 @@ static xi_state_t init_xively_topics( xi_context_handle_t in_context_handle );
-			XI_MQTT_RETAIN_FALSE, NULL, NULL );
+    char out_msg[IO_AXES_JSON_BUFFER_MAX_SIZE] = "";
+    SensorAxes_t sensor_input;
+
+    if ( io_read_accelero( &sensor_input ) < 0 )
+    {
+        printf( "\r\n\t[ERROR] trying to read accelerometer input" );
+        return -1;
+    } if ( io_axes_to_json( sensor_input, out_msg, IO_AXES_JSON_BUFFER_MAX_SIZE ) < 0 )
+    {
+        printf( "\r\n\t[ERROR] trying to create JSON string from sensor input" );
+        return -1;
+    }
+    return xi_publish( gXivelyContextHandle, mqtt_topic_descr->name, out_msg,
+                       XI_MQTT_QOS_AT_MOST_ONCE, XI_MQTT_RETAIN_FALSE, NULL, NULL );
 }
 
-xi_state_t pub_barometer( const mqtt_topic_descr_t* const mqtt_topic_descr )
+xi_state_t pub_magnetometer( const mqtt_topic_descr_t* const mqtt_topic_descr )
 {
-	return xi_publish( gXivelyContextHandle,
-			mqtt_topic_descr->name, "greetings from barometer",
-			XI_MQTT_QOS_AT_MOST_ONCE,
-			XI_MQTT_RETAIN_FALSE, NULL, NULL );
+    char out_msg[IO_AXES_JSON_BUFFER_MAX_SIZE] = "";
+    SensorAxes_t sensor_input;
+
+    if ( io_read_magneto( &sensor_input ) < 0 )
+    {
+        printf( "\r\n\t[ERROR] trying to read magnetometer input" );
+        return -1;
+    }
+    if ( io_axes_to_json( sensor_input, out_msg, IO_AXES_JSON_BUFFER_MAX_SIZE ) < 0 )
+    {
+        printf( "\r\n\t[ERROR] trying to create JSON string from sensor input" );
+        return -1;
+    }
+    return xi_publish( gXivelyContextHandle, mqtt_topic_descr->name, out_msg,
+                       XI_MQTT_QOS_AT_MOST_ONCE, XI_MQTT_RETAIN_FALSE, NULL, NULL );
 }
 
-xi_state_t pub_button( const mqtt_topic_descr_t* const mqtt_topic_descr )
+xi_state_t pub_gyroscope( const mqtt_topic_descr_t* const mqtt_topic_descr )
 {
-	return xi_publish( gXivelyContextHandle,
-			mqtt_topic_descr->name, "greetings from button",
-			XI_MQTT_QOS_AT_MOST_ONCE,
-			XI_MQTT_RETAIN_FALSE, NULL, NULL );
+    char out_msg[IO_AXES_JSON_BUFFER_MAX_SIZE] = "";
+    SensorAxes_t sensor_input;
+
+    if ( io_read_gyro( &sensor_input ) < 0 )
+    {
+        printf( "\r\n\t[ERROR] trying to read gyroscope input" );
+        return -1;
+    }
+    if ( io_axes_to_json( sensor_input, out_msg, IO_AXES_JSON_BUFFER_MAX_SIZE ) < 0 )
+    {
+        printf( "\r\n\t[ERROR] trying to create JSON string from sensor input" );
+        return -1;
+    }
+    return xi_publish( gXivelyContextHandle, mqtt_topic_descr->name, out_msg,
+                       XI_MQTT_QOS_AT_MOST_ONCE, XI_MQTT_RETAIN_FALSE, NULL, NULL );
 }
 
-xi_state_t pub_gyroscope( const mqtt_topic_descr_t* const mqtt_topic_descr )
+xi_state_t pub_barometer( const mqtt_topic_descr_t* const mqtt_topic_descr )
 {
-	return xi_publish( gXivelyContextHandle,
-			mqtt_topic_descr->name, "greetings from gyroscope",
-			XI_MQTT_QOS_AT_MOST_ONCE,
-			XI_MQTT_RETAIN_FALSE, NULL, NULL );
+    char out_msg[IO_FLOAT_BUFFER_MAX_SIZE] = "";
+    float sensor_input                     = 0;
+    io_read_pressure( &sensor_input );

```

Comment:
I'll get to that soon; this was just the first version of a working demo. I still need to add some error handling, clean up main.py and fix some edge cases.